cpp/roaring64map.hh

```diff
@@ -1098,90 +1100,69 @@ public:
-                            return true;
-                        },
-                        (void *)&map_entry.first);
-                });
-        } else
-            std::printf("{");
-        std::printf("}\n");
+        auto sink = [](const char *s) {
+            fputs(s, stdout);
+        };
+        printToSink(sink);
+        sink("\n");
     }
 
     /**
-     * Print the content of the bitmap into a string
+     * Print the contents of the bitmap into a string.
      */
     std::string toString() const {
-        struct iter_data {
-            std::string str{}; // The empty constructor silences warnings from pedantic static analyzers.
-            uint32_t high_bits{0};
-            char first_char{'{'};
-        } outer_iter_data;
-        if (!isEmpty()) {
-            auto map_iter = roarings.cbegin();
-            while (map_iter->second.isEmpty()) ++map_iter;
-            outer_iter_data.high_bits = roarings.begin()->first;
-            map_iter->second.iterate(
-                [](uint32_t low_bits, void *inner_iter_data) -> bool {
-                    ((iter_data *)inner_iter_data)->str +=
-                        ((iter_data *)inner_iter_data)->first_char;
-                    ((iter_data *)inner_iter_data)->str += std::to_string(
-                        uniteBytes(((iter_data *)inner_iter_data)->high_bits,
-                                   low_bits));
-                    ((iter_data *)inner_iter_data)->first_char = ',';
-                    return true;
-                },
-                (void *)&outer_iter_data);
-            std::for_each(
-                ++map_iter, roarings.cend(),
-                [&outer_iter_data](
-                    const std::pair<const uint32_t, Roaring> &map_entry) {
-                    outer_iter_data.high_bits = map_entry.first;
-                    map_entry.second.iterate(
-                        [](uint32_t low_bits, void *inner_iter_data) -> bool {
-                            ((iter_data *)inner_iter_data)->str +=
-                                ((iter_data *)inner_iter_data)->first_char;
-                            ((iter_data *)inner_iter_data)->str +=
-                                std::to_string(uniteBytes(
-                                    ((iter_data *)inner_iter_data)->high_bits,
-                                    low_bits));
-                            return true;
-                        },
-                        (void *)&outer_iter_data);
-                });
-        } else
-            outer_iter_data.str = '{';
-        outer_iter_data.str += '}';
-        return outer_iter_data.str;
+        std::string result;
+        auto sink = [&result](const char *s) {
+            result += s;
+        };
+        printToSink(sink);
+        return result;
+    }
+
+    /**
+     * Prints the contents of the bitmap to a caller-provided sink function.
+     */
+    void printToSink(const std::function<void(const char *)> &sink) const {
```

Comment:
IMO, if we're going to offer a public string-based API, we should use `std::string`. This relieves us and the user from caring about buffers and lengths.

Reply:
I think `printToSink` should be private. Both `toString()` and `printf()` are obviously fine. Extending the API with a `printToSink` function (which will require documentation and maintenance) should not be done *unless* it has been thoroughly discussed... Otherwise, we might be purchasing technical debt. I am not sure what the need is, here... If we are motivated by performance concerns, then we need benchmarks. If we don't care about performance, then the function we have are probably fine.

Reply:
The motivation of this PR is strictly source code readability, not performance. The two existing functions (`toString` and `printf`) are:
- similar but not the same
- both twice as long as they need to be
- hard to comprehend (in my opinion).

A lot of not-very-handsome code that ultimately does something very simple.

My motivation was to simplify and factor out their common functionality into a helper method. And we can keep the helper method private for now.

May I propose that I incorporate your feedback and do one more pass? Then you can take a look at the final product and make a decision whether to take it or drop it.

Reply:
OK all feedback addressed.

Old code was (by my count) 85 lines.
New code does the same thing in 48 lines and, to my eye, is more understandable.

Kindly let me know what you think.